### ISSUE TITLE
Running with msvs2017 too slow the first time.

### ISSUE THREAD

**Issue by joe2018Outlookcom** [timniederhausen/gn-build#15]
Is there anyone find that this build-gn config would run cost about 20 seconds, while running first time.
![image](https://user-images.githubusercontent.com/42833718/79219754-6e082880-7e85-11ea-8791-a2ece9809d12.png)

With the help of `gn help --tracelog` and `Chrome://tracing`，I found `\\build\toolchain\win\toolchain.py` cost most of time in that process.

I had a idea to verify the guess, so I rewrote the `\\build\toolchain\win\toolchain.py`， finally make it(Just a fake print).
![image](https://user-images.githubusercontent.com/42833718/79219801-811af880-7e85-11ea-9170-f3131e8ee3a3.png)

Then I figured out that, multiple processes were executed concurrently during its execution, which took too much time due to the slow running of `VC\Auxiliary\Build\vcvars32.bat`.
Compared with other compilers such as vs2013, the execution time of `VC\vcvarsall.bat` is very fast and the time consumed is very short.

I this way, I found that just run `VC\Auxiliary\Build\vcvars32.bat` once is enough, in the latest chromium’s  build-gn config. But this implemention should introduce some python scripts: `\\build\toolchain\win\setup_toolchain.py` and `\\build\vs_toolchain.py`. Besides, some modifies needed.

Is there anyone else agree with the idea?

**Comment by timniederhausen** [comment]
20 seconds is certainly too long, and not something I've experienced so far.
During local and CI builds for VS 2017 `gn gen` takes about 3 seconds, which is acceptable.

> I this way, I found that just run VC\Auxiliary\Build\vcvars32.bat once is enough

To support cross-compilation for different architectures, we need to run `vcvarsall.bat` for all supported target CPUs (x86, amd64, arm, arm64). To make this faster, all invocations happen in parallel.
Chromium too, needs to run their `setup_toolchain.py` for multiple architectures, albeit they only do so for similar architectures (i.e. x86/amd64 and arm/arm64)

> But this implemention should introduce some python scripts: 

These files contain a lot of chromium-specific stuff (support for their custom VS distribution etc.), which is not needed here. All required functionality is in `toolchain.py`

Can you tell me how long `gn gen` takes if you use Chromium's `//build`?
Additionally, have you tried cutting down the number of CPUs in [`toolchain.py`](https://github.com/timniederhausen/gn-build/blob/8ac02ced957264072940140e466e472b77298646/toolchain/win/toolchain.py#L277), as well as disabling [UWP](https://github.com/timniederhausen/gn-build/blob/8ac02ced957264072940140e466e472b77298646/toolchain/win/toolchain.py#L292)?

**Comment by joe2018Outlookcom** [comment]
> Can you tell me how long `gn gen` takes if you use Chromium's `//build`?

Not now, for some reasons(You konw why).

> Additionally, have you tried cutting down the number of CPUs in [`toolchain.py`](https://github.com/timniederhausen/gn-build/blob/8ac02ced957264072940140e466e472b77298646/toolchain/win/toolchain.py#L277), as well as disabling [UWP](https://github.com/timniederhausen/gn-build/blob/8ac02ced957264072940140e466e472b77298646/toolchain/win/toolchain.py#L292)?

Also, I assume that no manual changes to the script are needed. but I could try.



Through experiments, the time-consuming is related to the number of CPU cores.

(Image：Win10 1809，VS 15.9.21)

|  Core number   | time_1(ms)  | time_2(ms)  | time_3(ms)  | time_4(ms)  | time_5(ms)  | average_time(ms)  |
|  ----  | ----  | ----  | ----  | ----  | ----  | ----  |
| 1  | 23501 | 21977 | 17719 | 20126 | 17801 | 20225 |
| 2  | 16348 | 11684 | 11603 | 11666 | 11902 | 12641 |
| 4  | 6342 | 6577 | 6403 | 7244 | 7481 | 6809 |

So, I am sure it’s a problem instead of a normal phenomenon. 
I wonder if you were interestring of improving the problem?

**Comment by timniederhausen** [comment]
> Through experiments, the time-consuming is related to the number of CPU cores.

That's to be expected since all `vcsvarsall.bat` invocations happen in parallel.

I'll look into caching the results of `vcvarsall.bat`, but that won't help for the initial `gn gen`.
A workaround might be to limit the number of architectures supported by default, but I'd rather not do that.

There seems to be an [issue](https://developercommunity.visualstudio.com/content/problem/51179/vsdevcmdbat-or-vcvarsallbat-excecution-takes-a-ver.html) about long `vcvarsall.bat` execution times.
Perhaps the official fix makes any such measures obsolete, at least for VS 2019+

**Comment by joe2018Outlookcom** [comment]
Have there been any developments?.
